cluster/dht: Perform migrate-file with lk-owner

pranithk opened the pull request:
1) Added GF_ASSERT() calls in client-xlator to find these
issues sooner.
2) Fuse is setting zero-lkowner with len as 8 when the fop
doesn't have any lk-owner. Changed this to have len as 0
just as we have in fops triggered from xlators lower to
fuse.

fixes: #1529
Change-Id: Ia803db6b00869316893abb1cf435b898eec31228
Signed-off-by: Pranith Kumar K <pranith.karampuri@phonepe.com>

pranithk committed:
cluster/dht: Perform migrate-file with lk-owner

1) Added GF_ASSERT() calls in client-xlator to find these
issues sooner.
2) Fuse is setting zero-lkowner with len as 8 when the fop
doesn't have any lk-owner. Changed this to have len as 0
just as we have in fops triggered from xlators lower to
fuse.

fixes: #1529
Change-Id: Ia803db6b00869316893abb1cf435b898eec31228
Signed-off-by: Pranith Kumar K <pranith.karampuri@phonepe.com>

xhernandez commented on xlators/cluster/dht/src/dht-rebalance.c:
```diff
@@ -2359,17 +2360,32 @@ rebalance_task_completion(int op_ret, call_frame_t *sync_frame, void *data)
     int ret = -1;
+    call_frame_t *syncop_frame = NULL;
+
+    syncop_frame = copy_frame(frame);
```
Why is needed this copy ? the `frame` will exist until `DHT_STACK_UNWIND()` is called inside `rebalance_task_completion()`

pranithk replied:
@xhernandez Thought it is not a good idea to modify pid/lk-owner of frame that is coming from parent xlators. So had to go for copy_frame(). Do let me know if there is a way out. I couldn't think of any.

pranithk replied:
@xhernandez One thing we could probably do is to introduce another api into syncop which takes lk-owner and pid maybe others as arguments to set them to the op-frame. That is one option. Do you think that can be better?

pranithk replied:
I found another place where we can prevent a frame creation. Will update the patch.

pranithk replied:
@xhernandez Added a new commit.

xhernandez replied:
I think you are right. A new frame in this case seems ok.

pranithk committed:
syncop: Avoid frame allocation if we can

fixes: #1529
Change-Id: Iac3c7a00ee004fc7b7b4633e027171609150fea4
Signed-off-by: Pranith Kumar K <pranith.karampuri@phonepe.com>

pranithk commented:
Could other reviewers also take a look at this once please? This is a regression and should be fixed soon in all branches.

mohit84 reviewed (commented):
LGTM

pranithk committed:
cluster/dht: Set lkowner in daemon rebalance code path

fixes: #1529
Change-Id: Ibdb772b7010409f415d91551cddbfb36f44138af
Signed-off-by: Pranith Kumar K <pranith.karampuri@phonepe.com>

pranithk commented:
Had to force push to rebase the changes. Also found that rebalance is not setting lkowner. So fixed that part as well. I kept the commit log same as the previous one and only added a new commit to make it easier for reviews.

pranithk commented:
/run regression

pranithk committed:
cluster/afr: Set lkowner for ta-selfheal

fixes: #1529
Change-Id: Ic631c4b9d4e950efe82d365db69d2ec4b2a65781
Signed-off-by: Pranith Kumar K <pranith.karampuri@phonepe.com>

pranithk commented:
/run regression

pranithk committed:
cluster/ec: Destroy frame after heal is done

fixes: #1529
Change-Id: I5bb92ac438d9aa4c16293143be2bd595bb100a7a
Signed-off-by: Pranith Kumar K <pranith.karampuri@phonepe.com>

pranithk commented:
/run regression

pranithk commented:
/run regression

pranithk commented:
/run regression

pranithk commented:
I will handle lk asserts in client xlator and its dependencies into a separate issue and patch.

pranithk committed:
Don't assert for lk-owner in lk call

We need answers for #1704 before this can be tackled

Fixes: #1529
Change-Id: I96659d4fa9c888d368b58327df09c6c31dc4cc92
Signed-off-by: Pranith Kumar K <pranith.karampuri@phonepe.com>

pranithk commented:
/run regression

pranithk committed:
set lkowner for mandatory lock heal tests

Change-Id: Ia55cd54e2ebc67ba9391a9b80329b13147bd8022
Signed-off-by: Pranith Kumar K <pranith.karampuri@phonepe.com>

pranithk commented:
/run regression

pranithk commented:
Regressions have passed. Let me know if there are any other changes required. I will send one more patch for _lk changes based on responses from maintainers to my mail

BarakSason commented:
@pranithk 
I'd be happy to review, however I don't have enough knowledge regarding this in order to give a meaningful input.
Could you please describe more in details what's the original problem and how does these changes solve it?

pranithk commented:
> @pranithk
> I'd be happy to review, however I don't have enough knowledge regarding this in order to give a meaningful input.
> Could you please describe more in details what's the original problem and how does these changes solve it?

Generally any [f]inodelk/[f]entrylk operations are supposed to be sent with an lk-owner. If that is not set locks xlator unwinds with EINVAL. In this particular code path lk-owner is not set for the frame with which we are launching synctask. This patch fixes that part. I added asserts to see if there are other code paths which do the same. I found some more and fixed them too.

pranithk commented:
Could someone merge this patch if there are no more comments?
